### ISSUE TITLE
nngcat server cannot support more than 1.2K nngcat clients

### ISSUE THREAD

**Issue by tomghuang** [nanomsg/nng#1626]
**Describe the bug**

Since I cannot find any limitation on the amount of clients that an NNG server can support, so I assume that there is not limit. The only limit would be the resource of the machine.

To prove the theory, first, I launched the following nngcat server on one Ubuntu machine:
```
$ nngcat --rep --listen=tcp://0.0.0.0:55995 --data="42" --quoted
```

Then I use a Python script to launch 64K of the following long-running nngcat clients on another Ubuntu machine:
```
$ nngcat --req --dial=tcp://192.168.1.71:55995 --data="what is the answer?" --quoted --interval 60
```

**Expected behavior**

All 64K nngcat clients should successfully connect to the nngcat-based server.

**Actual Behavior**

After creating about 1.2K nngcat clients, we started to get the following error messages from the clients:
```
Unable to dial on tcp://192.168.1.71:55995: Connection refused
Unable to dial on tcp://192.168.1.71:55995: Timed out
```

**To Reproduce**

First, launch the nngcat server as mentioned above, and then use the following Python script to create 64K nngcat clients:
```python
import subprocess

p = []
for i in range(1024 * 64):
    ps = subprocess.Popen(['nngcat', '--req', '--dial=tcp://192.168.1.71:55995', '--data="what is the answer?"', '--quoted', '--interval', '60'],
                          shell=False,
                          stdout=subprocess.PIPE)
    p.append(ps)
    print(f'{i}', end='\r', flush=True)

p[0].communicate()
```

** Environment Details **
 - NNG version: 1.5.2
 - Operating system and version: Ubuntu 22.04
 - Compiler and language used: GCC and C
 - Shared or static library: static

**Additional context**

May I know if an NNG-based WebSocket server can support at least 64K WebSocket clients? Thanks.

**Comment by gdamore** [comment]
You're probably going to run into operating system limits:

1. Number of open files
2. Number of available ephemeral TCP ports

You can usually tune these, and its possible to raise the first to a much larger value (like 64K), but the second is usually harder.  There are typically only around 16K ephemeral ports available, although you can increase this effectively by using multiple IP address (each IP address gets its own pool of TCP ports).

I have run NNG with huge numbers of clients, and it does work with well over 10K.   I ran out of machine resource before I hit any predetermined limits.

**Comment by gdamore** [comment]
Btw, having 1200 nngcat clients is asking a *LOT* of your machine, because each of those is a whole process. This is a *MUCH* greater limit to scalability.

Generally speaking most systems cannot handle that many active processes.

**Comment by tomghuang** [comment]
@gdamore : Thank you very much for confirming that NNG/Mangos should be able to handle a large amount of clients.

Since nngcat might be too heavy-weight for this testing, I will implement a client for this purpose. I've tried to create 64K `nng_socket` and 64K corresponding dialers to simulated 64K clients. However, so far I can only create 28K sockets+dialers in a single machine. My next step is running the server in one machine, and running clients in other machines, and see if I can successfully prove that NNG/Mangos server can support at least 64K clients.

**Comment by gdamore** [comment]
The 28K is probably because of TCP ports.  Try using UNIX domain sockets. :-) 

Or configure multiple IP addresses on the machine.  You can for example try configureing 127.0.0.2, 127.0.0.3, etc.  and then using 10K clients from one, 10K clients from another, etc.  That *should* work.  You'll have to call listen in your server for each IP address, but that should be fine.

**Comment by tomghuang** [comment]
@gdamore : Really appreciate this multiple-IP trick. I'll try.

The reason why I need to use TCP socket or WebSocket for such testing, not UNIX domain socket, is because I'm trying to persuade a company using NNG/Mangos for their next generation IoT management system. They need to use cloud services to manage millions of devices deployed around the world, so I need to make sure that each NNG/Mangos-based server can at least handle 64K WebSocket clients.

**Comment by gdamore** [comment]
Cool.  NNG and mangos can both do this.  With fanout you can easily scale to really large networks -- millions of machines.

**Comment by tomghuang** [comment]
@gdamore : I've figured out how to create multiple IP addresses in a Ubuntu machine. However, after reading the NNG documentation, I still cannot find a way to bind a client `nng_socket` to a specific IP address.

I've tried to change `nng_sockaddr_in` for the client IP address after getting `nng_socket` by calling `nng_req0_open`, but I cannot use `nng_setopt` to change `nng_sockaddr_in`. In the `nng_options(5)` page, I don't see any way to change the client IP address.

Could you please give me a hint about how to specify the client IP address of a client NNG socket? Thanks.

**Comment by gdamore** [comment]
You need to create an nng_listener with a specific IP:

https://nng.nanomsg.org/man/tip/nng_listener_create.3.html

You can supply a URL like this: "tcp://127.0.0.3:5556"

**Comment by tomghuang** [comment]
@gdamore : Thanks for the suggestion, although I still didn't know how to make it work. I created a listener and a dialer for the same socket. I repeated this process 16K times for one client IP, and repeated 16K times for another client IP. But I still got "Address invalid" from dialers when 28K connections were established.

However, I finally figured out how to test if NNG can support at least 64K connections in a single machine with Docker. First, I launched the server in the host (docker IP: 172.17.0.1). Then I launched five containers, and each container run a client that would create 16K connections to the server. Since each container has its own IP (172.17.0.2, 172.17.0.3, 172.17.0.4, 172.17.0.5, 172.17.0.6), all the five clients could establish connections without any error. In the end, 16K x 5 = 80K connections were established.

With this technique, I should be able to prove that an NNG-based server can easily support a million connections. Thank you very much for making NNG so powerful. I can finally persuade the company to use NNG/Mangos.
